### PR TITLE
urandom-seed, urngd: avoid PKG_NAME in define lines

### PR DESCRIPTION
**Pull request by mans0n**
> Avoid reuse of PKG_NAME in call, define and eval lines for consistency and
> readability. Write the full name instead.

Ref: https://openwrt.org/docs/guide-developer/packages

While at it, remove urandom-seed definition from base-files Makefile.

### CODE REVIEW

**Comment by BKPepe** [comment]
I'd say that urandom-seed Makefile is still not okayish if you want to do polishing.

- There is an [deprecated SPDX License Identifier](https://spdx.org/licenses/).
- Build dir is wrong as there is no reason to overwrite defaults. Currently, it does not take care of PKG_VERSION, which is not a good thing. Take a look at [package.mk](https://github.com/openwrt/openwrt/blob/master/include/package.mk#L14).
- Use HTTPS instead of HTTP in `URL:=`.

All of this can be found in that reference, which you posted.

It that case following rows can be omitted
```
define Build/Prepare
	mkdir -p $(PKG_BUILD_DIR)
endef
```

And this one can be improved from
```
define Build/Compile/Default
endef
Build/Compile = $(Build/Compile/Default)
```

to:
```
Build/Compile:=:
```

Btw, maintainer could be set, but that's up author of that Makefile. 

So the result would be as there is need to increase `PKG_RELEASE` due to correctness in `PKG_LICENSE` and improvement in `URL:=`
```
include $(TOPDIR)/rules.mk

PKG_NAME:=urandom-seed
PKG_VERSION:=1.0
PKG_RELEASE:=2

PKG_MAINTAINER:=
PKG_LICENSE:=GPL-2.0-or-later

include $(INCLUDE_DIR)/package.mk

define Package/urandom-seed
  SECTION:=base
  CATEGORY:=Base system
  DEPENDS:=+getrandom
  TITLE:=/etc/urandom.seed handling for OpenWrt
  URL:=https://openwrt.org/
endef

Build/Compile:=:

define Package/urandom-seed/install
        $(CP) ./files/* $(1)/
endef

$(eval $(call BuildPackage,urandom-seed))
```

Compile tested and run tested by me.

**Comment by mans0n** [comment]
@BKPepe Wow, so much thanks for your detailed review!

Your patch might be applicable to other packages as well. I'll take care of it if possible.
Come to think of it, I don't feel PKG_VERSION is necessary here. What do you think?

**Comment by BKPepe** [comment]
Why do you think that `PKG_VERSION` should be removed? In my opinion, is that almost every package should have `PKG_VERSION`. It might have multiple reasons. 

- One of them can be just to know the difference between versions. 
For example, in version 1.0, there's an initial version, but in version 1.1/2.0 (depends on point of view), I have decided to bring new features and refactor and if you would remove it, `PKG_RELEASE` is not sufficient. For every change, you would need to increase the number of PKG_RELEASE, which will lead to hundreds, thousands and you can not reset it anytime. 

- There is also a possibility that you might change configure args to allow some features for OpenWrt, which is disabled by default in that package in upstream. So, it does not make sense to release a new version, right? Let's put aside that you don't have permission for that. For that, `PKG_RELEASE` is there. If there's a typo/changes in init scripts then `PKG_RELEASE` must be increased.  You wouldn't release a new version, because of that, right?

But this is said in [the documentation](https://openwrt.org/docs/guide-developer/packages) or in [CONTRIBUTING.md](https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md) in Packages repository.

And why in `PKG_BUILD_DIR` should be `PKG_VERSION`? In case, you wouldn't do clean.
You want to have it extracted to `PKG_NAME-PKG_VERSION`. When it extracts it just to `PKG_NAME` folder, you don't know which version was extracted there. In theory, you can check one package, but not tens, hundreds. As I said, there is no need to replace defaults with wrongs things.

I hope that I have answered all of it. If you have any other questions, don't hesitate to ask.

**Comment by mans0n** [comment]
urandom-seed was splited off from base-files, which has no PKG_VERSION.
I thought PKG_RELEASE is enough for packages which don't have their own repo (like package/system/mtd).
But I see your point, and I agree that we don't need to replace things by something that is not particularly better.
I'll just go with the one you suggested. Could I add your Tested-by?

**Comment by BKPepe** [comment]
Sure.

**Comment by adrianschmutzler** [comment]
Side-note: Since it's only two packages, please separate the commits for urandom-seed and urngd.

**Comment by mans0n** [comment]
Splited the PKG_NAME commit into two, and added urandom-seed clean-up patch.

@BKPepe Please have a look.

**Comment by mans0n** [comment]
After applying the patch, `make package/system/urandom-seed/dist V=s` returns error, but I think no one will care about it.

**Comment by adrianschmutzler** [comment]
Thanks, I've picked and applied the PKG_NAME patches for now.

**Comment by adrianschmutzler** [comment]
> Come to think of it, I don't feel PKG_VERSION is necessary here.

As I understand it, PKG_VERSION is not necessary here. PKG_VERSION is supposed to be used when you pull in external content, where the PKG_VERSION tracks "their" version numbers, while PKG_RELEASE tracks our revisions to it until PKG_VERSION increases.

> For example, in version 1.0, there's an initial version, but in version 1.1/2.0 (depends on point of view), I have decided to bring new features and refactor and if you would remove it, PKG_RELEASE is not sufficient. For every change, you would need to increase the number of PKG_RELEASE, which will lead to hundreds, thousands and you can not reset it anytime.

Indeed, that's how it is handled in base-files for example. For every change it just gets increased. If this works for relatively frequently changed base-files, I do not see why it should not work here.

For reference:
https://openwrt.org/docs/guide-developer/packages

"PKG_VERSION - The upstream version number that we're downloading"

(Note the word "upstream" ...)

File: package/system/urandom-seed/Makefile
Comment (adrianschmutzler) [inline package/system/urandom-seed/Makefile]: If you change the license, you need explicit Acked-by of all previous authors.

Comment (mans0n) [reply package/system/urandom-seed/Makefile]: Hmm, it's just a renaming, it refers the same license IMO.
The only contributor of urandom-seed Makefile is ynezz, but I'm not sure if we should include the list of contributors of base-files here...

**Comment by mans0n** [comment]
> So, I'd remove PKG_VERSION here. Again, it might be a good idea to send the reformatting of the package Makefile to the list, so ynezz as the "author" might have a chance to look at it and respond if he objects.

Okay, that seems a good compromise. I'll send a patch with PKG_VERSION dropped, and then ynezz can decide on it. I've also build and run tested, upgrade from version "1.0-1" to "2" succeeded without any error message.

**Comment by BKPepe** [comment]
> PKG_VERSION is not necessary here.

It's your opinion and I think it's wrong. Packages in folder `system` has some kind of versioning, which is done by `PKG_SOURCE_DATE` except zrap-swap and some other packages.

> relatively frequently changed base-files, I do not see why it should not work here.

Let's do the same mistake again. Right? It shouldn't harm as we can throw it away from other packages as well. :-) I'm wondering why there couldn't be `PKG_VERSION:=1.0`. It is not wrong.
However, this depends on author as well the maintainership if he wants to use versioning or not. He should use it.

> "PKG_VERSION - The upstream version number that we're downloading"

What is said in documentation is a holy bible and just one word can change it all. Just taking one word "upstream" is not enough. You can replace the word "upstream" to "major version". 

If you want to refer to another website, which was not mentioned you can take a look at:
https://openwrt.org/docs/guide-developer/overview

> package makefiles have PKG_VERSION that shows the upstream version (major version)`

There isn't anything about downloading files if we are trying nitpicking. See? There isn't any might/can. I don't know the reason, why `PKG_VERSION` can not be used. You can take it as doing a major version as said.

> Again, it might be a good idea to send the reformatting of the package Makefile to the list... > author" might have a chance to look at it...

In the first place, the package was already reviewed according to what you said, right? I don't know, why don't you use mention on github by @ and ynezz is active on Github as well, so let's wait till he respond. There was a weekend. I don't think it is necessary to respond immediately. On the other hand, you shouldn't force contributors to your mailing list instead of anything else as this is a valid way. Looking to https://openwrt.org/submitting-patches and as well there was a discussion here https://github.com/openwrt/openwrt/pull/1950 and with your [own Gitlab instance](https://lists.infradead.org/pipermail/openwrt-adm/2020-February/001322.html), which is being planned, it will be worse if you didn't review things on it.

File: package/system/urandom-seed/Makefile
Comment (BKPepe) [reply package/system/urandom-seed/Makefile]: @adrianschmutzler Where is the `LICENSE` file, hm? So it can be used variable `PKG_LICENSE_FILES`. Otherwise as said earlier current "GPL-2.0" is wrong.
Just take a look on [SPDX License list](https://spdx.org/licenses/).

**Review by BKPepe — APPROVED** [review]

File: package/system/urandom-seed/Makefile
Comment (mans0n) [reply package/system/urandom-seed/Makefile]: Oh, I was wrong. The list says that newer version of GPL-2.0 is GPL-2.0-only. Sorry for my confusion.

**Comment by mans0n** [comment]
I checked packages with source in the openwrt repo, and here is the list of packages that only use PKG_RELEASE:
boot/rbcfg
firmware/ipq-wifi
kernel/button-hotplug
kernel/gpio-button-hotplug
kernel/gpio-nct5104d
kernel/hwmon-gsc
kernel/i2c-gpio-custom
kernel/leds-apu2
kernel/rtc-rv5c386a
kernel/spi-gpio-custom
kernel/w1-gpio-custom
libs/gettext
libs/libiconv
system/mtd
utils/fbtest
utils/fritz-tools
utils/jboot-tools
utils/nvram
utils/osafeloader
utils/oseama
utils/otrx
utils/px5g
utils/rbextract
utils/usbreset

And this is the list of packages which do have PKG_VERSION:
kernel/om-watchdog
kernel/trelay
system/urandom-seed
system/zram-swap

I'm not sure this is in the former or the latter:
utils/ct-bugcheck

I still think that using only PKG_RELEASE is enough for those packages, but if I'm wrong, it might not be easy to revert the change...

Aside from the version and maintainer, the last patch does the following:
- Fix deprecated SPDX license id
- change `http://openwrt.org/` to `https://openwrt.org/`
- drop unnecessary custom PKG_BUILD_DIR and Build/Prepare
- simplify Build/Compile definition

All these can be applicable to other packages, and I feel some needs of drawing more feedback on the changes. Now I think of it, each of the changes can go into separate commits. If I write those patches, I'll try sending them to the mailing list. For this PR, I'll wait for review from @ynezz here.

**Comment by mans0n** [comment]
Set PKG_LICENSE to GPL-2.0-only. Also dropped blank PKG_MAINTAINER for now.

**Comment by adrianschmutzler** [comment]
"PKG_VERSION - The upstream version number that we're downloading"

This says the it's the "upstream version". Full stop. There is no room for interpretation to it.
So, I do not have to justify why I want to follow that rule but we should argument about why to _deviate_ from it.

The system as used in the guidelines has advantages: Upstream versions (PKG_VERSION, PKG_SOURCE_DATE) are clearly separated from "our" version (PKG_RELEASE). That's easy to describe and easy to understand for everyone. You will always know which version to bump.

If we deviated from that now, I only see disadvantages:
- For every bump, you would have to consider whether it's a major or minor bump, and which version you would have to change. This is especially problematic for a rarely changed package without a maintainer ...
- If you use PKG_VERSION for versioning, and at some point a package goes upstream, it might happen that it would start with 1.0 there, while you locally might already have 2.1 and would have to work around the version decrease somehow.
- If we really want to have major.minor versioning, the whole version string would belong into PKG_RELEASE. There you could do a 1.2, 1.3.2, ... then without breaking the concept of upstream version vs. "our" version.

> package makefiles have PKG_VERSION that shows the upstream version (major version)`

This essentially says the same: "package makefiles have PKG_VERSION that shows the upstream version" means PKG_VERSION is upstream version. "(major version)" just is more precise on which upstream version is to be used.

> On the other hand, you shouldn't force contributors to your mailing list instead of anything else as this is a valid way.

See, that's where I'm pragmatic. First, I'm not forcing anybody. But it is known that there much less people on GitHub compared to the list. I do not think it's fair to keep people here waiting if I know they might have quicker or wider feedback on the list. So, I suggest to submit to the list in cases where I think the list might give them more attention.
For the discussion at hand, as I've expressed above, the current state of this package as well as your opinion deviates from what is described in the guide lines. So, I think it might be best to discuss it with a big audience, and maybe settle it in general, rather than having the two of us (or three) going in circles here. If so, we might only have this discussion once and won't have to repeat it next time.

Despite that, please be aware that I welcome your feedback on tidying up this package in the first place. It's not like I couldn't just enforce my view by merging the version I want instead of discussing about it.

**Comment by adrianschmutzler** [comment]
@ynezz Since this is based on one of your commits, can you have a look? Should be a quick one for you to review and merge (you may just ignore our PKG_VERSION discussion here) ...

File: package/system/urandom-seed/Makefile
Comment (adrianschmutzler) [inline package/system/urandom-seed/Makefile:21]: I haven't checked the flags, but it seems that the default Build/Prepare does different/additional things?

File: package/system/urandom-seed/Makefile
Comment (adrianschmutzler) [inline package/system/urandom-seed/Makefile:18]: What's that? I didn't find any other statement like this in package directory.

**Comment by adrianschmutzler** [comment]
I will merge the first remaining patch later today.

For the second one, it might be easier to just drop the changes on the define blocks and I'd merge the rest.
However, if you want to clean up more substantially, you are also welcome to discuss this.

File: package/system/urandom-seed/Makefile
Comment (mans0n) [reply package/system/urandom-seed/Makefile:18]: This line nullify [Build/Compile/Default](https://github.com/openwrt/openwrt/blob/master/include/package-defaults.mk#L144). AFAICT `:` is same as `true` in shell (it does nothing). Probably we can just use [Build/Compile=](https://github.com/openwrt/openwrt/blob/master/package/system/openwrt-keyring/Makefile#L31) here.

File: package/system/urandom-seed/Makefile
Comment (mans0n) [reply package/system/urandom-seed/Makefile:21]: IIRC `Build/Prepare/Default` does nothing harmful, and `mkdir -p $(PKG_BUILD_DIR)` is executed somewhere in the general package initialization process. But I can leave it if you like.

**Comment by adrianschmutzler** [comment]
@neheb May you can resolve this quickly, as you should be more experienced in this matter:
Do the Build/Prepare / Build/Compile changes here make sense or should we leave that untouched?

File: package/system/urandom-seed/Makefile
Comment (jow-) [reply package/system/urandom-seed/Makefile:18]: Personally I like the explicit empty define notation because it is immediately clear to a casual reader that it is an empty section. Could you live with keeping it like it was?

**Comment by neheb** [comment]
Second @jow- 's comments. Empty define sections are more common.

**Comment by mans0n** [comment]
Thanks for the reviews. I dropped the define blocks changes.

**Comment by adrianschmutzler** [comment]
Thanks, merged.

**Pull request closed without merging**
